### PR TITLE
feature: upgrade Astro to v5 beta

### LINKED ISSUE #188: Upgrade to Astro v5
[Astro v5 (beta)](https://astro.build/blog/astro-5-beta/) is out! We should upgrade and test if everything works. There's quite a bit of changes, so good to follow the [Astro v5 upgrade guide](https://5-0-0-beta.docs.astro.build/en/guides/upgrade-to/v5/).

### PR DESCRIPTION
**Pull request by luukbrauckmann**
# Changes

- Upgrade Astro to the version 5 beta

# Associated issue

Resolves #188

# How to test

1. Open preview link
2. See if it works
3. Check code mainly

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have made sure that my PR is easy to review (not too big, includes comments)
- ~~I have made updated relevant documentation files (in project README, docs/, etc)~~
- ~~I have added a decision log entry if the change affects the architecture or changes a significant technology~~
- [x] I have notified a reviewer

### CODE REVIEW

File: src/lib/datocms.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ type DatocmsRequest = {
  */
 export const datocmsRequest = async <T>({ query, variables = {}, retryCount = 1 }: DatocmsRequest): Promise<T> => {
   const headers = new Headers({
-    Authorization: getSecret('DATOCMS_READONLY_API_TOKEN'),
+    Authorization: getSecret('DATOCMS_READONLY_API_TOKEN')!,
```
Comment (luukbrauckmann) [inline src/lib/datocms.ts:23]: `getSecret('DATOCMS_READONLY_API_TOKEN')` is possibly `undefined`, but I think it should always be set right? If it wouldn't be set then you wouldn't be able to get data.

File: src/middleware.ts [hunk 2]
```diff
@@ -15,8 +15,10 @@ export const hashSecret = async (secret: string) => {
 };
 
 export const datocms = defineMiddleware(async ({ locals }, next) => {
-  locals.datocmsEnvironment = datocmsEnvironment;
-  locals.datocmsToken = getSecret('DATOCMS_READONLY_API_TOKEN');
+  Object.assign(locals, {
```
Comment (luukbrauckmann) [inline src/middleware.ts:18]: From the Astro docs. Should set new locals through `Object.assign()`

File: tsconfig.json [hunk 3]
```diff
@@ -19,5 +19,6 @@
       }
     ]
   },
-  "exclude": ["node_modules", "dist"]
+  "include": [".astro/types.d.ts", "**/*"],
```
Comment (luukbrauckmann) [inline tsconfig.json:22]: Astro types can now be included in the `tsconfig.json`

File: astro.config.ts [hunk 4]
```diff
@@ -53,7 +51,7 @@ export default defineConfig({
     service: passthroughImageService()
   },
   integrations: [sitemap()],
-  output: isPreview ? 'server' : 'hybrid',
+  output: isPreview ? 'server' : 'static',
```
Comment (luukbrauckmann) [inline astro.config.ts:54]: hybrid is merged with static. Static is the default now.

File: src/env.d.ts [hunk 5]
```diff
@@ -1,4 +1,3 @@
-/// <reference path="../.astro/env.d.ts" />
```
Comment (luukbrauckmann) [inline src/env.d.ts:1]: Not necessary anymore

File: package.json [hunk 6]
```diff
@@ -40,13 +40,13 @@
     "postinstall": "npx husky install"
   },
   "dependencies": {
-    "@astrojs/check": "^0.7.0",
-    "@astrojs/cloudflare": "^10.4.0",
-    "@astrojs/sitemap": "^3.0.5",
+    "@astrojs/check": "^0.9.3",
+    "@astrojs/cloudflare": "^12.0.0-beta.0",
```
Comment (jurgenbelien) [inline package.json:44]: Interested in the changes here

File: src/lib/datocms.ts (hunk 1 above)
Comment (jurgenbelien) [reply src/lib/datocms.ts:23]: What happens when it is not set? Does it throw a nice error that the environment variable is missing, or do I only see the authorization fail? As a user I would be best served with an error that I'm missing the `.env`.

File: src/middleware.ts (hunk 2 above)
Comment (jurgenbelien) [reply src/middleware.ts:18]: I see that it is intended as being immutable in page and middleware context. 

> Astro 5.0 requires you to append values to the existing locals object without deleting it. Locals in middleware, API endpoints, and pages, can no longer be completely overridden.
https://5-0-0-beta.docs.astro.build/en/guides/upgrade-to/v5/#changed-adding-values-to-contextlocals

**Review by jurgenbelien — COMMENTED** [review]
Looks good but should we keep this open as a `next` branch?

File: src/lib/datocms.ts (hunk 1 above)
Comment (luukbrauckmann) [reply src/lib/datocms.ts:23]: Yeah I agree, but I also thought that one of the first things you do is adding a `.env` file. The error I get now doesn't really tell you what's wrong. Maybe we should put a warning in a sort of `predev` script?

```
luuk@Luuks-MBP head-start % npm run dev

> head-start@0.0.1 predev
> run-s prep


> head-start@0.0.1 prep
> run-s prep:* --print-label

[prep:clean-astro-env      ]
[prep:clean-astro-env      ] > head-start@0.0.1 prep:clean-astro-env
[prep:clean-astro-env      ] > rm -rf node_modules/.astro
[prep:clean-astro-env      ]
[prep:cloudflare-env       ]
[prep:cloudflare-env       ] > head-start@0.0.1 prep:cloudflare-env
[prep:cloudflare-env       ] > touch .dev.vars
[prep:cloudflare-env       ]
[prep:datocms-types        ]
[prep:datocms-types        ] > head-start@0.0.1 prep:datocms-types
[prep:datocms-types        ] > graphql-codegen --config .graphqlrc.ts
[prep:datocms-types        ]
[prep:datocms-types        ] (node:35547) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
[prep:datocms-types        ] (Use `node --trace-deprecation ...` to show where the warning was created)
[prep:datocms-types        ] Saving generated types for DatoCMS (environment: 'page-model-names') to 'src/lib/types/datocms.d.ts'.
[prep:datocms-types        ] [STARTED] Parse Configuration
[prep:datocms-types        ] [SUCCESS] Parse Configuration
[prep:datocms-types        ] [STARTED] Generate outputs
[prep:datocms-types        ] [STARTED] Generate to src/lib/types/datocms.d.ts
[prep:datocms-types        ] [STARTED] Load GraphQL schemas
[prep:datocms-types        ] [FAILED] Could not obtain introspection result, received the following as response;
[prep:datocms-types        ] [FAILED]  { data: [{ id: "5413bc", type: "api_error", attributes: [Object] }] }
[prep:datocms-types        ] [SUCCESS] Generate outputs
ERROR: "prep:datocms-types" exited with 1.
ERROR: "prep" exited with 1.
luuk@Luuks-MBP head-start %
```

Comment (jbmoelker) [reply src/lib/datocms.ts:23]: `dotenv-safe` already safeguards the `prep` scripts. Those should be able to function standalone.

For the `astro.config.ts` we could do the same, and throw a meaningful error in there to ensure all env variables are defined and that we don't need to check elsewhere in the code.

File: src/middleware.ts (hunk 2 above)
Comment (jbmoelker) [reply src/middleware.ts:18]: `datocmsEnvironment: datocmsEnvironment,` can be simplified to `datocmsEnvironment`.

File: src/middleware.ts [hunk 7]
```diff
@@ -32,16 +34,18 @@ const i18n = defineMiddleware(async ({ params, request }, next) => {
...
 const preview = defineMiddleware(async ({ cookies, locals }, next) => {
-  const previewSecret = getSecret('HEAD_START_PREVIEW_SECRET');
-  locals.isPreview = getSecret('HEAD_START_PREVIEW');
-  locals.isPreviewAuthOk = Boolean(previewSecret) && cookies.get(previewCookieName)?.value === await hashSecret(previewSecret);
-  locals.previewSecret = previewSecret;
+  const previewSecret = getSecret('HEAD_START_PREVIEW_SECRET')!;
+  Object.assign(locals, {
+    isPreview: getSecret('HEAD_START_PREVIEW'),
+    isPreviewAuthOk: Boolean(previewSecret) && cookies.get(previewCookieName)?.value === await hashSecret(previewSecret),
+    previewSecret: previewSecret
```
Comment (jbmoelker) [inline src/middleware.ts:47]: `previewSecret: previewSecret` -> `previewSecret`

**Comment by jbmoelker** [comment]
Very happy with all the test coverage we have so far. Gives me a much better feeling with these upgrades 🙌

File: package.json (hunk 6 above)
Comment (jbmoelker) [reply package.json:44]: As far as I can tell ensures `astro:env` and correct `output` mode is supported (`hybrid` is now `static`).

**Review by jbmoelker — APPROVED** [review]
Approved! Everything seems to work as it should. Very happy with all the test coverage we have so far. Gives me a much better feeling with these upgrades 🙌

Added some small suggestions for improvements, but approved as is.

**Pull request merged**